pkg/metrics/controller.go

```diff
@@ -258,11 +271,60 @@ func (c *Controller) recordGameServerStatusChanges(old, next interface{}) {
 	if newGs.Status.State != oldGs.Status.State {
 		fleetName := newGs.Labels[agonesv1.FleetNameLabel]
 		if fleetName == "" {
-			fleetName = "none"
+			fleetName = defaultFleetTag
 		}
 		recordWithTags(context.Background(), []tag.Mutator{tag.Upsert(keyType, string(newGs.Status.State)),
 			tag.Upsert(keyFleetName, fleetName)}, gameServerTotalStats.M(1))
+
+		// Calculate the duration of the current state
+		duration, err := c.calcDuration(oldGs, newGs)
+		if err != nil {
+			c.logger.Debug(err.Error())
+		} else {
+			recordWithTags(context.Background(), []tag.Mutator{tag.Upsert(keyType, string(oldGs.Status.State)),
+				tag.Upsert(keyFleetName, fleetName)}, gsStateDurationSec.M(duration))
+		}
+	}
+}
+
+// calcDuration - calculate duration between state changes
+// store current time from creationTimestamp for each update received
+// Assumptions: there is a possibility that one of the previous state change timestamps would be evicted,
+// this measure would be skipped
+func (c *Controller) calcDuration(oldGs, newGs *agonesv1.GameServer) (duration float64, err error) {
+	duration = 0.
+	// currentTime - GameServer time from its start
+	currentTime := time.Now().UTC().Sub(newGs.ObjectMeta.CreationTimestamp.Local().UTC()).Seconds()
+
+	newGSKey := fmt.Sprintf("%s/%s/%s", newGs.ObjectMeta.Namespace, newGs.ObjectMeta.Name, string(newGs.Status.State))
+	oldGSKey := fmt.Sprintf("%s/%s/%s", oldGs.ObjectMeta.Namespace, oldGs.ObjectMeta.Name, string(oldGs.Status.State))
+
+	c.stateLock.Lock()
+	defer c.stateLock.Unlock()
+	switch {
+	case newGs.Status.State == agonesv1.GameServerStateCreating || newGs.Status.State == agonesv1.GameServerStatePortAllocation:
+		duration = currentTime
+	case !c.gameServerStateLastChange.Contains(oldGSKey):
+		err = errors.New(fmt.Sprintf("Unable to calculate '%s' state duration of '%s' GameServer", oldGs.Status.State, oldGs.ObjectMeta.Name))
+	default:
+		val, ok := c.gameServerStateLastChange.Get(oldGSKey)
+		if !ok {
+			err = errors.New(fmt.Sprintf("Could not find expected key %s", oldGSKey))
```

Comment:
> Error strings should not be capitalized (unless beginning with proper nouns or acronyms) or end with punctuation, since they are usually printed following other context.

```suggestion
			err = errors.New(fmt.Sprintf("could not find expected key %s", oldGSKey))
```